src/CommandSet.hpp

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <functional>
+#include <map>
+#include <string>
+#include <utility>
+#include "string_utils.hpp"
+
+template<typename T>
+class CommandSet {
+ public:
+  CommandSet() {}
+  
+  // add a command called 'name' with a value, which is available to characters
+  // at level 'level' or above.
+  void add(const std::string& name_in, T value, int level) {
+    auto name = lower_case(name_in);
+    commands_[name] = std::make_pair(value, level);
+  }
+  
+  std::optional<const T> get(const std::string& name_in, int level) const {
+    auto name = lower_case(name_in);
+    // lower_bound gets the lowest item whose key is great than or equal to name.
+    for (auto it = commands_.lower_bound(name); it != commands_.end(); it++) {
```

Comment:
There's a cute thing for this: [equal_range](https://en.cppreference.com/w/cpp/container/map/equal_range) returns a pair of iterators that bound the range. _Maybe_ useful here?

```
auto &&[first, last] = commands_.equal_range(name);
for (auto it = first; it != last; ++it)
```

I'm not 100% sure that's the same as what you're doing though!

Reply:
From what I can see, that's just a way of getting lower_bound and upper_bound together. Whereas I want entries that the query is a prefix of (which would be outside the 'equal range'). With the proviso that I only match stuff which is available to the player's level. So it doesn't look like equal_range is really what I want. I guess it's intended for std::multimap etc.